![image](competi.png)

### En esta primera prueba hago un encode y busco scores de algoritmo polinómico individuales para cada columna con price.

#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

In [44]:
import numpy as np
import pandas as pd

import urllib.request
from PIL import Image

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score
from sklearn import preprocessing



In [45]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
df = pd.read_csv("diamonds_train.csv", index_col=0)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40345 entries, 0 to 40344
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    40345 non-null  float64
 1   cut      40345 non-null  object 
 2   color    40345 non-null  object 
 3   clarity  40345 non-null  object 
 4   depth    40345 non-null  float64
 5   table    40345 non-null  float64
 6   x        40345 non-null  float64
 7   y        40345 non-null  float64
 8   z        40345 non-null  float64
 9   price    40345 non-null  int64  
dtypes: float64(6), int64(1), object(3)
memory usage: 3.4+ MB


In [4]:
df.head(20)

carat        cut color clarity  depth  table     x     y     z  price
0    1.21      Ideal     H     VS2   63.0   57.0  6.73  6.70  4.23   6134
1    0.28  Very Good     D    VVS2   64.0   56.0  4.14  4.17  2.66    532
2    0.42    Premium     F     VS1   61.2   58.0  4.86  4.82  2.96   1103
3    0.26      Ideal     H      IF   61.1   57.0  4.16  4.12  2.53    600
4    1.10       Good     G     SI1   63.4   57.0  6.52  6.55  4.14   4997
5    0.35      Ideal     F     SI1   62.2   56.0  4.48  4.52  2.80    583
6    0.31      Ideal     I     VS2   62.3   54.0  4.37  4.40  2.73    452
7    0.43    Premium     H     SI1   60.2   57.0  4.93  4.91  2.96    919
8    0.51    Premium     I     SI2   60.4   59.0  5.20  5.17  3.13    956
9    1.22      Ideal     I     VS1   61.9   55.7  6.81  6.84  4.23   6469
10   0.41  Very Good     D     SI1   63.2   57.0  4.77  4.73  3.00    947
11   0.33    Premium     G    VVS1   61.9   59.0  4.41  4.38  2.72   1114
12   0.93      Ideal     I     SI1   62.5   55.0  6.23  6.29  3.91   3856
13   0.71      Ideal     H     VS1   61.6   57.0  5.74  5.79  3.55   2175
14   0.47      Ideal     G    VVS2   62.2   53.0  5.00  5.04  3.12   1618
15   1.71    Premium     I     SI2   60.6   57.0  7.78  7.71  4.69   9760
16   0.33  Very Good     E    VVS2   60.1   63.0  4.45  4.50  2.69    981
17   1.26  Very Good     G     SI1   61.7   60.0  6.85  6.90  4.24   6733
18   0.75    Premium     G     SI1   60.9   58.0  5.87  5.83  3.56   2631
19   0.76  Very Good     H     SI1   61.4   57.0  5.89  5.94  3.63   2557

In [47]:
le = preprocessing.LabelEncoder()

print(df["cut"].unique())
print(df["color"].unique())
print(df["clarity"].unique())


need_label = ["cut", "color", "clarity"]

for column in need_label:
    data = le.fit_transform(df[column])
    df[column] = data
    print(df.head())

['Ideal' 'Very Good' 'Premium' 'Good' 'Fair']
['H' 'D' 'F' 'G' 'I' 'E' 'J']
['VS2' 'VVS2' 'VS1' 'IF' 'SI1' 'SI2' 'VVS1' 'I1']
   carat  cut color clarity  depth  table     x     y     z  price
0   1.21    2     H     VS2   63.0   57.0  6.73  6.70  4.23   6134
1   0.28    4     D    VVS2   64.0   56.0  4.14  4.17  2.66    532
2   0.42    3     F     VS1   61.2   58.0  4.86  4.82  2.96   1103
3   0.26    2     H      IF   61.1   57.0  4.16  4.12  2.53    600
4   1.10    1     G     SI1   63.4   57.0  6.52  6.55  4.14   4997
   carat  cut  color clarity  depth  table     x     y     z  price
0   1.21    2      4     VS2   63.0   57.0  6.73  6.70  4.23   6134
1   0.28    4      0    VVS2   64.0   56.0  4.14  4.17  2.66    532
2   0.42    3      2     VS1   61.2   58.0  4.86  4.82  2.96   1103
3   0.26    2      4      IF   61.1   57.0  4.16  4.12  2.53    600
4   1.10    1      3     SI1   63.4   57.0  6.52  6.55  4.14   4997
   carat  cut  color  clarity  depth  table     x     y     z  p

**Importante:**

   - Si quitas columnas o creas columnas nuevas a partir de otras, o cualquier modificación column-wise tendrás que aplicarlo al dataset de `diamonds_test.csv` de cara a hacer la predicción.
   - Si por lo contrario, decides por ejemplo, quitar los outliers o hacer un `dropna()`, o cualquier modificación row-wise eso NO PODRÁS (ni debes) aplicarlo al dataset de `diamonds_test.csv` de cara a hacer la predicción. ¿Por qué? Porque si el conjunto de test tiene 50 observaciones (filas) la predicción se espera que tenga 50 filas.

#### ¿Y si lo que hago es estandarizar los datos o hacer un encoding, también se lo tengo que hacer al test antes de predecir?

In [ ]:
# ¿qué opináis?
#Sí, habría que hacerlo

-----------------------------------------------------------------------------------------------------------------

### Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [48]:
# coger las columnas "x", "y", "z" tiene solo valor ilustrativo, NO LO COPIÉIS
l_X = df[["x","y","z", "carat", "depth", "table", "cut", "color", "clarity"]].values
l_X = np.transpose(l_X)
y = df["price"].values
l_X.shape

(9, 40345)

In [49]:
y.shape

(40345,)

#### ¿Y si después del análisis exploratorio de mis datos llego a la conclusión de que puedo predecir el precio solo con las columnas `x`, `y` y `z`, también aplica al test?

In [ ]:
# ¿qué opináis?
# sí, el conjunto para predecir `diamonds_test.csv` también tendrá que tener 3 columnas

### 2. Dividir X_train, X_test, y_train, y_test

In [8]:
# el test_size 0.20 y el random_state puede cambiar
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

NameError: name 'X' is not defined

In [8]:
X_train

array([[6.5 , 6.55, 4.16],
       [5.34, 5.38, 3.31],
       [4.41, 4.38, 2.7 ],
       ...,
       [6.05, 6.07, 3.73],
       [6.5 , 6.53, 3.99],
       [6.51, 6.56, 4.1 ]])

### 3. Asignar el modelo (vacío) a una variable
Aquí meteríais los parámetros. 

**Consejo**: Usa GridSearch y vuélvete loca o loco probando modelos e hiperparámetros.

In [9]:
model = LinearRegression()

Siempre hay tiempo para una historia:
https://catboost.ai/docs/concepts/python-reference_catboostregressor.html

### 4. Entrenar el modelo con X_train, y_train

In [50]:
lin_reg_model = LinearRegression()
for i, X in enumerate(l_X):
    
    X = X.reshape(-1, 1)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)
    print("\n")
    tabla_score = []
    tabla_x = []
    for degree in range(2,40):
        polinominal_model = PolynomialFeatures(degree) 
        X_poly = polinominal_model.fit_transform(X_train.reshape(-1, 1), y_train)
        X_poly_T = polinominal_model.fit_transform(X_test.reshape(-1, 1), y_test)
        print(X_poly_T.shape)
        print(y_test.shape)
        lin_reg_model.fit(X_poly, y_train)
        

        y_pred = lin_reg_model.predict(X_poly)
        tabla_score.append(r2_score(y_train, y_pred)*100)
        tabla_x.append(degree)
        print (f'Score con "{degree}" grados de dato {i} para Price:', r2_score(y_train, y_pred)*100)
        
        predictions = lin_reg_model.predict(X_poly_T)
        np.sqrt(mean_squared_error(y_test, predictions))
        print("MSE:", np.sqrt(mean_squared_error(y_test, predictions)))




(8069, 3)
(8069,)
Score con "2" grados de dato 0 para Price: 86.18626888110983
MSE: 1517.291002053524
(8069, 4)
(8069,)
Score con "3" grados de dato 0 para Price: 86.22582121270969
MSE: 1518.902715474148
(8069, 5)
(8069,)
Score con "4" grados de dato 0 para Price: 86.63342239243602
MSE: 1465.1663415433568
(8069, 6)
(8069,)
Score con "5" grados de dato 0 para Price: 86.89938080407408
MSE: 1447.0703683791787
(8069, 7)
(8069,)
Score con "6" grados de dato 0 para Price: 86.90283197418016
MSE: 1446.4925291085592
(8069, 8)
(8069,)
Score con "7" grados de dato 0 para Price: 86.93562132699837
MSE: 1443.0742680541912
(8069, 9)
(8069,)
Score con "8" grados de dato 0 para Price: 86.95526917131068
MSE: 1440.7447124833504
(8069, 10)
(8069,)
Score con "9" grados de dato 0 para Price: 86.95549839509087
MSE: 1440.7003289171248
(8069, 11)
(8069,)
Score con "10" grados de dato 0 para Price: 86.96634676772568
MSE: 1439.8196119451006
(8069, 12)
(8069,)
Score con "11" grados de dato 0 para Price: 86.9700

In [26]:
X = np.array(df[["x","y","z"]])
X.shape

(40345, 3)

In [27]:
y = np.array(df["price"])
y.shape

(40345,)

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)
print("\n")
for degree in range(2,3):
    polinominal_model = PolynomialFeatures(degree) 
    print(X_train.shape)
    X_poly = polinominal_model.fit_transform(X_train.reshape(-1, 1))
    #X_poly = polinominal_model.fit_transform(X_train.reshape(-1, 1), y_train)
    lin_reg_model = LinearRegression()
    print(X_poly.shape)
    X_poly_T = polinominal_model.fit_transform(X_test.reshape(-1, 1), y_test)
    print(X_poly_T.shape)
    print(y_test.shape)
    lin_reg_model.fit(X_poly, y_train)
        

    y_pred = lin_reg_model.predict(X_poly)
    tabla_score.append(r2_score(y_train, y_pred)*100)
    tabla_x.append(degree)
    print (f'Score con "{degree}" grados para Price:', r2_score(y_train, y_pred)*100)
        
    predictions = lin_reg_model.predict(X_poly_T)
    np.sqrt(mean_squared_error(y_test, predictions))
    print("MSE:", np.sqrt(mean_squared_error(y_test, predictions)))



(32276, 3)
(96828, 3)
(24207, 3)
(8069,)


ValueError: Found input variables with inconsistent numbers of samples: [96828, 32276]

In [10]:
model.fit(X_train, y_train)

LinearRegression()

### 5. Predecir con el modelo ya entrenado con X_test

In [11]:
predictions = model.predict(X_test)
print(predictions)

[ 7704.85364418  5221.32255796  -458.15129534 ... -1757.15918457
  6025.68092069  2426.18738372]


### 6. Sacar métricas, valorar el modelo 

Recuerdo que en la competición se va a evaluar con la métrica de RMSE.

In [12]:
np.sqrt(mean_squared_error(y_test, predictions))

1894.008376646408

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

_Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `diamonds_test.csv` nos arriesgamos a que digamos, no salga lo esperado._

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `diamonds_train.csv` al completo.

**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**

Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `diamonds_test.csv` para predecir.

**¿De dónde saco `diamonds_test.csv`?**

In [17]:
X_pred = pd.read_csv("diamonds_test.csv", index_col = 0)
X_pred.head()

carat        cut color clarity  depth  table     x     y     z
0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

In [18]:
X_pred.shape

(13449, 9)

**¿Por qué da error?** 

In [19]:
predictions_submit = model.predict(X_pred)
predictions_submit

ValueError: could not convert string to float: 'Ideal'

## IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 13449 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `diamonds_test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [20]:
X_pred = pd.read_csv("diamonds_test.csv", index_col=0)
X_pred.head()

carat        cut color clarity  depth  table     x     y     z
0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

**¿Qué más habrá que quitar o hacer?**

In [ ]:
# ¿Qué opináis?
# En este caso, quitar todas las columnas, menos la x, y y z

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**Truqui**: Carga X que debe estar actualmente actualizada a cómo la usaste para entrenar a `model` y haz que `X_pred` sea igual. ¡Sin quitar filas!

In [21]:
X

array([[6.73, 6.7 , 4.23],
       [4.14, 4.17, 2.66],
       [4.86, 4.82, 2.96],
       ...,
       [5.33, 5.34, 3.32],
       [6.47, 6.56, 3.88],
       [5.33, 5.27, 3.2 ]])

In [22]:
X_pred = np.array(X_pred[['x','y','z']])

Porque:
    - SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 3 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS

In [23]:
X_pred

array([[4.41, 4.43, 2.65],
       [4.52, 4.46, 2.79],
       [7.58, 7.55, 4.56],
       ...,
       [7.3 , 7.37, 4.53],
       [7.79, 7.68, 4.66],
       [4.29, 4.34, 2.64]])

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [ ]:
# respuestas

In [24]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([ -260.20029487,   116.70396517,  9699.26794807, ...,
        8903.59284445, 10336.93013348,  -601.00685048])

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

**¿De dónde saco `sample_submission.csv`?**

In [25]:
sample = pd.read_csv("sample_submission.csv")

In [26]:
sample.head()

id  price
0   0  12132
1   1  11786
2   2  14684
3   3  15425
4   4   6724

In [27]:
sample.shape

(13449, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [28]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "price": predictions_submit})

In [29]:
submission.head()

id        price
0   0  -260.200295
1   1   116.703965
2   2  9699.267948
3   3  -496.020397
4   4  8831.773091

In [30]:
submission.shape

(13449, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [31]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [32]:
help(chequeator)

Help on function chequeator in module __main__:

chequeator(df_to_submit)
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.



In [34]:
chequeator(submission)

You're ready to submit!
